src/main/flight/alt_hold.c

```diff
@@ -0,0 +1,197 @@
+    simplePid_s altPid;
+    simplePid_s velPid;
+    float throttle;
+    float targetAltitude;
+    float measuredAltitude;
+    float measuredAccel;
+    float velocityEstimationAccel;  // based on acceleration
+    float startVelocityEstimationAccel;
+    bool prevAltHoldModeEnabled;
+} altHoldState_s;
+
+void altHoldReset(altHoldState_s* altHoldState)
+{
+    simplePidInit(&altHoldState->altPid, -50.0f, 50.0f,
+                  0.01f * altholdConfig()->altPidP,
+                  0.0f,
+                  0.01f * altholdConfig()->altPidI);
+
+    simplePidInit(&altHoldState->velPid, 0.0f, 1.0f,
+                  0.01f * altholdConfig()->velPidP,
+                  0.01f * altholdConfig()->velPidD,
+                  0.0f);
+
+    altHoldState->throttle = 0.0f;
+    altHoldState->startVelocityEstimationAccel = altHoldState->velocityEstimationAccel;
+    altHoldState->targetAltitude = (float)(0.01f * getEstimatedAltitudeCm());
+}
+
+void altHoldInit(altHoldState_s* altHoldState)
+{
+    altHoldState->prevAltHoldModeEnabled = false;
+    altHoldState->velocityEstimationAccel = 0.0f;
+    altHoldReset(altHoldState);
+}
+
+
+void altHoldUpdate(altHoldState_s* altHoldState)
+{
+    bool altHoldModeEnabled = FLIGHT_MODE(ALTHOLD_MODE);
+
+    if (altHoldModeEnabled && !altHoldState->prevAltHoldModeEnabled)
+    {
+        altHoldReset(altHoldState);
+    }
+    altHoldState->prevAltHoldModeEnabled = altHoldModeEnabled;
+
+
+    float measuredAltitude = (float)(0.01f * getEstimatedAltitudeCm());
+
+    t_fp_vector accelerationVector = {{
+        acc.accADC[X],
+        acc.accADC[Y],
+        acc.accADC[Z]
+    }};
+
+    imuTransformVectorBodyToEarth(&accelerationVector);
+
+    float measuredAccel = 9.8f * (accelerationVector.V.Z - acc.dev.acc_1G) / acc.dev.acc_1G;
+
+    DEBUG_SET(DEBUG_ALTHOLD, 0, (int16_t)(measuredAccel * 100.0f));
+
+    altHoldState->velocityEstimationAccel += measuredAccel * 0.01f;
+    altHoldState->velocityEstimationAccel *= 0.999f;
+
+    float currentVelocityEstimationAccel = altHoldState->velocityEstimationAccel - altHoldState->startVelocityEstimationAccel;
+    DEBUG_SET(DEBUG_ALTHOLD, 1, (int16_t)(100.0f * currentVelocityEstimationAccel));
+
+    altHoldState->measuredAltitude = measuredAltitude;
+    altHoldState->measuredAccel = measuredAccel;
+
+    float velocityTarget = simplePidCalculate(&altHoldState->altPid, 0.01f, altHoldState->targetAltitude, altHoldState->measuredAltitude);
+    DEBUG_SET(DEBUG_ALTHOLD, 2, (int16_t)(100.0f * velocityTarget));
+
+    float velPidForce = simplePidCalculate(&altHoldState->velPid, 0.01f, velocityTarget, currentVelocityEstimationAccel);
+    
+    float newThrottle = velPidForce;
+
+    DEBUG_SET(DEBUG_ALTHOLD, 3, (int16_t)(100.0f * velPidForce));
+
+    newThrottle = constrainf(newThrottle, 0.0f, 1.0f);
+    newThrottle = scaleRangef(newThrottle, 0.0f, 1.0f, 0.01f * altholdConfig()->minThrottle, 0.01f * altholdConfig()->maxThrottle);
+
+    if (!altHoldModeEnabled) {
```

Comment:
This should be refactored - estimate vertical velocity unconditionally (ideally in IMU) and enable ALTHOLD pids only when necessary.

Reply:
It's easier to debug if everything calculated all the time (no need to plug battery and turn radio).
100hz light-weight calculation is nothing compared to 8K PID loops, no need to optimize anything here.